src/org/rascalmpl/dap/RascalDebugAdapter.java

```diff
@@ -548,6 +588,141 @@ public CompletableFuture<EvaluateResponse> evaluate(EvaluateArguments args) {
+                    detail.append(((IString) v).getValue());
+                    detail.append("\n");
+                }
+            }
+        }
+        return detail.toString();
+    }
+
+    private List<CompletionItem> getVariablesCompletionsFromFrame(IRascalFrame frame, String text) {
+        List<CompletionItem> completions = new ArrayList<>();
+        Set<String> frameVariables = frame.getFrameVariables();
+        for(String var : frameVariables){
+            if(var.startsWith(text)){
+                CompletionItem completion = new CompletionItem();
+                completion.setLabel(var);
+                completion.setSortText(var);
+                if (text.length() < var.length()) { //remove the prefix
+                    completion.setText(var.substring(text.length()));
+                }
+                else { // exact match
+                    completion.setText("");
+                }
+                completion.setType(CompletionItemType.VARIABLE);
+                completions.add(completion);
+            }
+        }
+        return completions;
+    }
+
+    private List<CompletionItem> getFunctionsCompletionsFromEnvironment(Environment env, String text) {
+        List<CompletionItem> completions = new ArrayList<>();
+        for(Pair<String,LinkedHashSet<AbstractFunction>> functions : env.getFunctions()){
+            String funcName = functions.getFirst();
+            if(funcName.startsWith(text)){
+                for(AbstractFunction func : functions.getSecond()){
+                    CompletionItem completion = new CompletionItem();
+                    completion.setLabel(func.getHeader());
+                    completion.setSortText(funcName);
+                    if (text.length() < funcName.length()) { //remove the prefix
+                        completion.setText(funcName.substring(text.length()));
+                    }
+                    else { // exact match
+                        completion.setText("");
+                    }
+                    completion.setType(CompletionItemType.FUNCTION);
+                    completion.setDetail(getFunctionDetail(func));
+                    completions.add(completion);
+                }
+            }
+        }
+        return completions;
+    }
+
+    @Override
+    public CompletableFuture<CompletionsResponse> completions(CompletionsArguments args) {
+        return CompletableFuture.supplyAsync(() -> {
+            CompletionsResponse response = new CompletionsResponse();
+            List<CompletionItem> completions = new ArrayList<>();
+            if(suspendedState.isSuspended()) {
+                Integer frameId = args.getFrameId();
+                assert frameId != null;
+                IRascalFrame frame = suspendedState.getCurrentStackFrames()[frameId];
+
+                // Variable names starting with the typed text
+                completions.addAll(getVariablesCompletionsFromFrame(frame, args.getText()));
```

Comment:
How about we pass along the `completions` instance, such that we're not writing to intermediate array list just to concat them here? (same for the other functions).